rllib/agents/trainer.py

```diff
@@ -914,19 +952,32 @@ def step_attempt(self) -> ResultDict:
                 with concurrent.futures.ThreadPoolExecutor() as executor:
                     train_future = executor.submit(
                         lambda: next(self.train_exec_impl))
-                    if self.config["evaluation_num_episodes"] == "auto":
-
-                        # Run at least one `evaluate()` (num_episodes_done
-                        # must be > 0), even if the training is very fast.
-                        def episodes_left_fn(num_episodes_done):
-                            if num_episodes_done > 0 and \
+                    # Automatically determine duration of the evaluation.
+                    if self.config["evaluation_duration"] == "auto":
+                        unit = self.config["evaluation_duration_unit"]
+
+                        # Run at least one `evaluate()`, even if the training
+                        # is very fast.
+                        def duration_fn(remaining_duration):
+                            # Training is done and we already ran at least one
+                            # evaluation -> Nothing left to run.
+                            if remaining_duration > 0 and \
                                     train_future.done():
                                 return 0
-                            else:
+                            # Count by episodes. -> Run n more
+                            # (n=num eval workers).
+                            elif unit == "episodes":
                                 return self.config["evaluation_num_workers"]
```

Comment:
actually I don't quite get the logic here anymore. duration_fn gets called over and over to see whether the evaluation should stop, right? why do we always return self.config["evaluation_num_workers"] here, which is a positive number?
as written, eval only stops as soon as remaining_duration > 0 and train_future is done?

Reply:
Correct, it gets called over and over to determine whether there is some eval duration left (>0) to run.

For `evaluation_duration_unit=episodes` (and auto duration!):
- Run n more episodes: Each worker will - on a call to its `sample()` run exactly one episode (as this is configured via the worker's batch_mode=complete_episodes and rollout_fragment_length=1 settings).
- Then check again, whether training is done and we can stop

For `evaluation_duration_unit=timesteps` (and auto duration!):
- Run [num-workers * fragment-len * num_envs_per_worker] more timesteps: note that this is the minimum that one `sample()` on all workers will do.

In the non-auto case:
The result of the `duration_fn` is used to determine, how many more `sample()` (on which eval workers) we need to call. This is important to reach the exact desired number of timesteps/episodes with n workers.